### DISCUSSION TITLE
Weird behavior on AWS beanstalk deployment

### DISCUSSION THREAD

**Discussion by RodrigoNeves95** [miguelgrinberg/Flask-SocketIO#1948] [Q&A]
I'm kinda running out of ideas and getting desperate at the same time ahah. I also don't know if this is the correct place but here it goes:

Trying to deploy a flask web-app using flask-socketio over a Docker AWS Beanstalk deployment. gevent is being used to serve the web app:

Note 1: The deployment used an AWS Application load balancer and entry point with a certificate and routes the traffic to a EC2 instance where a nginx server is running on port 80 to get the request and proxy them to the docker on port 5000 - This is the normal flow of the AWS Beanstalk stack

Note: no configuration at all was made to the nginx on the EC2 instnace

```
CMD ["gunicorn", "--bind", "0.0.0.0:5000","--workers","1", "--threads", "1", "application:application", "-k", "geventwebsocket.gunicorn.workers.GeventWebSocketWorker", "--timeout", "180"]
```

First weird thing:

From the client if I do not force the protocol to to only use WebSocket I can't make the server go from polling to WebSocket. I get a 400 on the request. If I force to use Webscocket from the client I can make the ws protocol work.

Second weird thing:

If I actually remove gevent-websocket requirement and start the server with a gevent worker the server can make the request over polling, with no problem at all.

Third weird thing:

If I run the server with gevent-websocket support and force the client to run with ws only (to make sure I can actually use webscokets) I can run the web app normally, except for one tiny little thing that is driving me crazy: The first request to the server (after the login) is getting a 400 regardless of what I do. And this request is a normal HTTP GET method. All the requests after that first one, runs normally. 

I'm also getting this weird things on the server output

```
<gevent._socket3.socket at 0x7fd3c5d2c1c0 object, fd=14, family=2, type=1, proto=6>: Invalid HTTP method: '3ff5ce1-583e6c977c0b74f445ff044d\r\n' 
```

I have no idea what this is and where it comes from

Happy to present all info needed

**Comment by miguelgrinberg** [comment]
> I also don't know if this is the correct place but here it goes:

No. It's not good to file a bug when you just have questions. Use the discussions board for questions. I have transferred it for you this time.

> From the client if I do not force the protocol to to only use WebSocket I can't make the server go from polling to WebSocket. I get a 400 on the request. If I force to use Webscocket from the client I can make the ws protocol work.

This is expected when you use a load balancing solution that is not compliant with the requirements of Socket.IO. See 
[using multiple workers](https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-multiple-workers) in the documentation for these requirements.

> If I actually remove gevent-websocket requirement and start the server with a gevent worker the server can make the request over polling, with no problem at all.

I don't see what relation gevent-websocket vs gevent standard worker can have on this, so I don't have an explanation for this other than you changed something else without realizing it.

> If I run the server with gevent-websocket support and force the client to run with ws only (to make sure I can actually use webscokets) I can run the web app normally, except for one tiny little thing that is driving me crazy: The first request to the server (after the login) is getting a 400 regardless of what I do. And this request is a normal HTTP GET method. All the requests after that first one, runs normally.

I don't understand this. If you are connecting with WebSocket, then there is only one request per connection. No idea what other requests your client is sending, but if you are using WebSocket they cannot be related to Socket.IO.

**Reply by RodrigoNeves95** [reply]
Thanks for moving into a discussion, indeed is more natural making this a discussion!

>This is expected when you use a load-balancing solution that is not compliant with the requirements of Socket.IO. See
[using multiple workers](https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-multiple-workers) in the documentation for these requirements.

This would make sense to me, only if it were impossible to run the Websocket protocol. As I said when I force the client to connect through the websocket protocol only, everything just works. If the reverse proxy would not be able to handle WebSockets, then this would not be true, right?

> I don't see what relation gevent-websocket vs gevent standard worker can have on this, so I don't have an explanation for this other than you changed something else without realizing it.

Yeah me neither. What I know is that it works for some reason, is just weird. Probably, running it with gevent without gevent-webscoket makes the server only attempt to communicate over polling (since webscocket protocol is not supported without gevent-websocket)

> I don't understand this. If you are connecting with WebSocket, then there is only one request per connection. No idea what other requests your client is sending, but if you are using WebSocket they cannot be related to Socket.IO.

This is exactly what is driving me crazy here. Like does not make sense. I can just fake a get request that will fail with a 400 and then everything will just work. I have no really good explanation. I'm also not sure that the request is reaching the flask app as well. The only log that I get is that weird message error that I posted that does not seem very informative. I can't debug and reach where it comes from.

I know all of this is just weird and deploying this through the AWS services adds lot of configuration to the mix. I ended up writing the issue because, while not changing anything in the infra, removing the gevent-webscoket, and running with polling as the protocol I don't get any of this. 

What can change in terms of reverse proxy config by using gevent-websocket and not using it?

PS: Well kind of expected maybe, but worth saying, everything works locally

PPS: You mentioned the multiple workers I'm currently running a redis queue and using sticky sessions to make sure I can scale. Currently debugging with only one server running

**Reply by RodrigoNeves95** [reply]
Kept digging, and from what I can see it seems like the nginx reverse proxy can handle the socket-io protocol 
<img width="1282" alt="image" src="https://user-images.githubusercontent.com/24609535/222227932-59588708-5eb0-46b2-994b-adb58693a09c.png">

**Reply by miguelgrinberg** [reply]
> This would make sense to me, only if it were impossible to run the Websocket protocol. As I said when I force the client to connect through the websocket protocol only, everything just works.

This is actually the clue that tells me this problem is caused by incompatible load balancing. The 400s happen when several requests from a client are load balanced into different servers. If you switch to WebSocket there is a single request, so load balancing does not affect these connections.

> Kept digging, and from what I can see it seems like the nginx reverse proxy can handle the socket-io protocol

Correct, nginx has full support for Socket.IO, if you configured it with sticky sessions (example in the docs). But your top-level load balancer is not Nginx, right? You are using Elastic Beanstalk.

**Reply by RodrigoNeves95** [reply]
The deployment should have two reverse proxy. One, which is the entry point for AWS infra, comes from the Application Load balancer (created when AWS Beanstalk stack gets created) and we have the reverse proxy in the EC2 instance running the docker with the server. The image above is from this second proxy.

The sticky sessions are configured at the Application load balancer level, because this is where the route gets proxied to different targets (EC2 instances) where only one worker (gevevnt server) is running

By the way I already tested this behavior with auto scaling on and everything seems to work (apart from the initial problem raised with that first single request)

Trying to reproduce locally with similar nginx configuration (so far not able)

> This is actually the clue that tells me this problem is caused by incompatible load balancing. The 400s happen when several requests from a client are load balanced into different servers. If you switch to WebSocket there is a single request, so load balancing does not affect these connections.

I agree with you,  tried to force sticky sessions, even with only one instance running. No success going from polling to webscoket

**Reply by RodrigoNeves95** [reply]
Well small update. Check the following configuration for a local nginx deployment where I'm getting the same behavior

<img width="1274" alt="image" src="https://user-images.githubusercontent.com/24609535/222250574-c64f832f-f49c-41db-925f-78b714b52fdd.png">


Adding network tab with the behavior 
<img width="803" alt="image" src="https://user-images.githubusercontent.com/24609535/222251404-0a9a06d3-eb77-43ab-990e-d1a987c4ee08.png">
 
and console output 
<img width="634" alt="image" src="https://user-images.githubusercontent.com/24609535/222251449-020c6b3b-b565-4330-8d48-b3ea6905db7e.png">

**Reply by miguelgrinberg** [reply]
What do you have running on 127.0.0.1:5000?

**Reply by RodrigoNeves95** [reply]
Is the flask backend being served as following:

```
gunicorn --bind 127.0.0.1:5000 application:application -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker  --reload --timeout 1800 --threads 1
```

**Reply by RodrigoNeves95** [reply]
That weird gevent._socket3.socket appears everytime the 400 happens actually 
<img width="987" alt="image" src="https://user-images.githubusercontent.com/24609535/222252731-c7826df1-7d02-4e0a-91b0-cbe3913e6906.png">

**Reply by miguelgrinberg** [reply]
How many workers is gunicorn starting? I don't remember if the -w option is required to get one worker.

**Reply by RodrigoNeves95** [reply]
I have 99% sure is one, as you clearly point out in the docs. I will force the deployment to run with one worker with the -w flag and report back

**Reply by miguelgrinberg** [reply]
I'm afraid the 400 errors suggest you have more than one worker running. This would be easy to figure out from the logs Gunicorn writes during start up.

**Reply by RodrigoNeves95** [reply]
As I was expecting I never ran the server with more than one worker. Check the server output logs after starting explicitly with one worker and making a login and logout twice 
<img width="2978" alt="image" src="https://user-images.githubusercontent.com/24609535/222420944-37f6653e-b731-42ac-b352-cb166eeeb01d.png">
<img width="897" alt="image" src="https://user-images.githubusercontent.com/24609535/222421019-7d3ec018-bd74-42ff-bfab-764895152669.png">


That weird socket print appears every time we get the 400. 

I also tried locally not using gevent-websocket (running everything with polling, and not trying the upgrade) and the 400 does not happen. Something that you can remember from gevent-websocket that is making this happening?

**Reply by miguelgrinberg** [reply]
I'm sorry, I think I may have not been very clear. When I referred to the 400 errors I meant errors coming from the Socket.IO side. This is not a Socket.IO error, this is some issue on the gevent-websocket library, which I'm not surprised of since it hasn't been regularly maintained in a long time. You can try using eventlet, which has its own internal WebSocket server.

**Reply by RodrigoNeves95** [reply]
Thanks for the feedback and answers. I will give it a try and report back then.

This is not a blocker for me, because after the first 400 everything just works for the session, but nevertheless, I found this super weird.

Thanks for the support

**Reply by RodrigoNeves95** [reply]
Finally got the time to remove gevent + gevent-webscoket and use eventlet. 

Everything works. I had to do it because things got unusable. So I initially thought that we would only get an initial 400. But actually, I would get an 400 error everytime after a POST request (yeah I have no idea as well)

I will assume gevent-websocket is just broken.

**Reply by RodrigoNeves95** [reply]
I think we can close this
